### PR TITLE
[SYCL][L0] Force submit of dependent open command batches

### PR DESCRIPTION
**Pull request by smaslov-intel**
Previously we were only force-submitting batches seen as dependency in different PI queue.
Now, we also properly submit copy/compute open command-lists **within** the same PI queue too.

E2E test: https://github.com/alexanderfle/llvm-test-suite/blob/reuse_events_in_L0_tests/SYCL/Plugin/level_zero_inorder_interleaving_kernel_copy.cpp

Signed-off-by: Sergey V Maslov <sergey.v.maslov@intel.com>

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -1702,20 +1696,33 @@ pi_result _pi_ze_event_list_t::createAndRetainPiZeEventList(
...
+
+            if (Queue == CurQueue &&
+                OpenCommandList->second.isCopy(Queue) == UseCopyEngine) {
+              // Don't force execute the batch yet since the new command
+              // is going to the same open batch as the dependent event.
+            } else {
+              if (auto Res = Queue->executeOpenCommandList(
+                      OpenCommandList->second.isCopy(Queue)))
+                return Res;
+            }
```
Comment (againull) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1723]: From this code it looks like we always force execute open command list if it is compute command list. Could you please explain why  behavior needs to be different for copy and compute command lists? I think comment https://github.com/intel/llvm/pull/6199/files#diff-15dd1eb076d2164bd9e87d9057f05f652a716498e8cdf5975e564c65309a0985R1706-R1709 should mention about this difference.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1723]: It is not different for copy vs. compute. This code force-submits copy if the dependent command is compute, and vice versa. Perhaps, you were confused by this check `OpenCommandList->second.isCopy(Queue) == UseCopyEngine` looking like it is about copy-only, but I think it is not :)

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 2]
```diff
@@ -1702,20 +1696,33 @@ pi_result _pi_ze_event_list_t::createAndRetainPiZeEventList(
...
-
-          if (auto Res = Queue->executeAllOpenCommandLists())
-            return Res;
-        } else if (!Queue) {
+        if (Queue) {
+          // The caller of createAndRetainPiZeEventList must already hold
+          // a lock of the CurQueue. Additionally lock the Queue if it
+          // is different from CurQueue.
+          auto Lock = ((Queue == CurQueue) ? std::unique_lock<pi_shared_mutex>()
+                                           : std::unique_lock(Queue->Mutex));
```
[the remark this answers is not in the document]
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: I agree. Added TODO comment to rework this further

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 1 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1725]: Oh, I see, yes, I was confused by this check.

**Review by againull — APPROVED** [review]

**Pull request merged**